### PR TITLE
Polls and voting database and view model functionality

### PR DESCRIPTION
**Pull request by juaniac**
This PR is meant to:
- add a poll data class for an event poll which contains an id, a title, a query and a list of options that users can vote on and a map mapping users to the option they voted on.  
- add database functions to add / delete a poll, vote an option on a poll.
- add event view model functions to add / delete a poll, vote an option on a poll.
- add robust tests for all of the above functionality.

### CODE REVIEW

File: app/src/main/java/com/monkeyteam/chimpagne/viewmodels/EventViewModel.kt [hunk 1]
```diff
@@ -402,6 +407,65 @@ class EventViewModel(
         {})
   }
 
+  fun createAPollAtomically(
```
Comment (Athanaze) [inline app/src/main/java/com/monkeyteam/chimpagne/viewmodels/EventViewModel.kt:410]: As discussed previously, you should not put articles in function names

File: app/src/androidTest/java/com/monkeyteam/chimpagne/newtests/model/event/EventTest.kt [hunk 2]
```diff
@@ -125,4 +128,106 @@ class EventTest {
...
+
+    loading = true
+    eventManager.atomic.castPollVote(
+        event.id,
+        poll.id,
+        account2.firebaseAuthUID,
+        "2",
+        { loading = false },
+        { assertTrue(false) })
+    while (loading) {}
```
Comment (Athanaze) [inline app/src/androidTest/java/com/monkeyteam/chimpagne/newtests/model/event/EventTest.kt:190]: a helping function waitLoading would reduce copy pasted code here

File: app/src/androidTest/java/com/monkeyteam/chimpagne/newtests/model/event/EventTest.kt [hunk 3]
```diff
@@ -125,4 +128,106 @@ class EventTest {
...
+    updatedEvent = ChimpagneEvent()
+    loading = true
+    eventManager.getEventById(
+        event.id,
+        {
+          updatedEvent = it!!
+          loading = false
+        },
+        { assertTrue(false) })
+    while (loading) {}
```
Comment (Athanaze) [inline app/src/androidTest/java/com/monkeyteam/chimpagne/newtests/model/event/EventTest.kt:202]: here

File: app/src/androidTest/java/com/monkeyteam/chimpagne/newtests/model/event/EventTest.kt [hunk 4]
```diff
@@ -125,4 +128,106 @@ class EventTest {
...
+    assertEquals("1", updatedEvent.polls[poll.id]!!.votes[account0.firebaseAuthUID])
+    assertEquals("1", updatedEvent.polls[poll.id]!!.votes[account1.firebaseAuthUID])
+    assertEquals("2", updatedEvent.polls[poll.id]!!.votes[account2.firebaseAuthUID])
+
+    assertEquals(2, updatedEvent.polls[poll.id]!!.getNumberOfVotesPerOption()["1"])
+    assertEquals(1, updatedEvent.polls[poll.id]!!.getNumberOfVotesPerOption()["2"])
+
+    loading = true
+    eventManager.atomic.deletePoll(event.id, poll.id, { loading = false }, { assertTrue(false) })
+    while (loading) {}
```
Comment (Athanaze) [inline app/src/androidTest/java/com/monkeyteam/chimpagne/newtests/model/event/EventTest.kt:218]: here

File: app/src/androidTest/java/com/monkeyteam/chimpagne/newtests/model/event/EventTest.kt [hunk 5]
```diff
@@ -125,4 +128,106 @@ class EventTest {
...
+
+    loading = true
+    eventManager.getEventById(
+        event.id,
+        {
+          updatedEvent = it!!
+          loading = false
+        },
+        { assertTrue(false) })
+    while (loading) {}
```
Comment (Athanaze) [inline app/src/androidTest/java/com/monkeyteam/chimpagne/newtests/model/event/EventTest.kt:228]: and here

File: app/src/main/java/com/monkeyteam/chimpagne/viewmodels/EventViewModel.kt (hunk 1 above)
Comment (juaniac) [reply app/src/main/java/com/monkeyteam/chimpagne/viewmodels/EventViewModel.kt:410]: Indeed I did not follow the conventions, it is fixed now.

File: app/src/androidTest/java/com/monkeyteam/chimpagne/newtests/model/event/EventTest.kt (hunk 2 above)
Comment (juaniac) [reply app/src/androidTest/java/com/monkeyteam/chimpagne/newtests/model/event/EventTest.kt:190]: This cannot be done for database related functions as established.
Plus you would be replacing 1 line of code by another 1 line of code in the tests, so I do not see the point of trying to make modifications.

File: app/src/androidTest/java/com/monkeyteam/chimpagne/newtests/model/event/EventTest.kt (hunk 3 above)
Comment (juaniac) [reply app/src/androidTest/java/com/monkeyteam/chimpagne/newtests/model/event/EventTest.kt:202]: same as above

File: app/src/androidTest/java/com/monkeyteam/chimpagne/newtests/model/event/EventTest.kt (hunk 4 above)
Comment (juaniac) [reply app/src/androidTest/java/com/monkeyteam/chimpagne/newtests/model/event/EventTest.kt:218]: same as above

File: app/src/androidTest/java/com/monkeyteam/chimpagne/newtests/model/event/EventTest.kt (hunk 5 above)
Comment (juaniac) [reply app/src/androidTest/java/com/monkeyteam/chimpagne/newtests/model/event/EventTest.kt:228]: same as above.

File: app/src/androidTest/java/com/monkeyteam/chimpagne/newtests/model/event/EventTest.kt (hunk 2 above)
Comment (QuicheLorraine3600) [reply app/src/androidTest/java/com/monkeyteam/chimpagne/newtests/model/event/EventTest.kt:190]: > a helping function waitLoading would reduce copy pasted code here

It's hard to do has you cannot directly pass a pointer in Kotlin. You need to wrap loading into an object... Not sure we want to make our testing framework more complex

File: app/src/main/java/com/monkeyteam/chimpagne/model/database/AtomicChimpagneEventManager.kt [hunk 6]
```diff
@@ -112,4 +112,45 @@ class AtomicChimpagneEventManager(
         .addOnSuccessListener { onSuccess() }
         .addOnFailureListener(onFailure)
   }
+
+  fun createPoll(
```
Comment (QuicheLorraine3600) [inline app/src/main/java/com/monkeyteam/chimpagne/model/database/AtomicChimpagneEventManager.kt:116]: Note that if you pass an already existing poll, the old one would be overwritten. I personally prefer to name the method "updatePoll". However this name is also not perfect, as you might think a poll needs to exist to be updated.

File: app/src/main/java/com/monkeyteam/chimpagne/model/database/ChimpagnePoll.kt [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+
+data class ChimpagnePoll(
+    val id: ChimpagnePollId = UUID.randomUUID().toString(),
+    val title: String = "",
+    val query: String = "",
+    val options: Map<ChimpagnePollOptionId, ChimpagnePollOption> = emptyMap(),
+    val votes: Map<ChimpagneAccountUID, ChimpagnePollOptionId> = emptyMap(),
+) {
+  private fun getVotesPerOptions(): Map<ChimpagnePollOptionId, List<ChimpagneAccountUID>> {
+    val votesPerOptions: MutableMap<ChimpagnePollOptionId, List<ChimpagneAccountUID>> = ArrayMap()
```
Comment (QuicheLorraine3600) [inline app/src/main/java/com/monkeyteam/chimpagne/model/database/ChimpagnePoll.kt:18]: You've forgotten Martin Odersky's course :sob:

**Review by QuicheLorraine3600 — APPROVED** [review]
Really nice work. I think there are no concurrency issues with your atomic methods. 
You are free to address or not my comments, there are not really relevant for the rest of the project.
One last thing: please stop naming View Models "VM", it's really disturbing and you're the only one using this.

Merge whenever you want :+1:

File: app/src/main/java/com/monkeyteam/chimpagne/model/database/AtomicChimpagneEventManager.kt (hunk 6 above)
Comment (juaniac) [reply app/src/main/java/com/monkeyteam/chimpagne/model/database/AtomicChimpagneEventManager.kt:116]: Yes, but the intent of the function is to create a poll, and you cannot edit / update a poll once it has been created thus the situation you describe should never happen and is not the intended use case of this function.

File: app/src/main/java/com/monkeyteam/chimpagne/model/database/ChimpagnePoll.kt (hunk 7 above)
Comment (juaniac) [reply app/src/main/java/com/monkeyteam/chimpagne/model/database/ChimpagnePoll.kt]: Yes I broke protocol, but it is a sacrifice I willing to make.

**Pull request merged**